ujt/callbacks.py

```diff
@@ -0,0 +1,188 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#    http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Callbacks for Dash app. """
+
+from collections import deque
+from typing import Dict, Tuple, cast
+
+import dash
+import dash_html_components as html
+from dash.dependencies import ALL, Input, Output, State
+from dash.exceptions import PreventUpdate
+from graph_structures_pb2 import Client, Node, NodeType
+
+from . import (
+    compute_status,
+    constants,
+    converters,
+    generate_data,
+    state,
+    transformers,
+    utils)
+from .dash_app import app
+
+
+@app.callback(
+    Output("cytoscape-graph",
+           "elements"),
+    [
+        Input("refresh-button",
+              "n_clicks_timestamp"),
+        Input({"datatable-id": ALL},
+              "selected_row_ids")
+    ],
+    [State("cytoscape-graph",
+           "elements")],
+)
+def update_graph_elements(n_clicks_timestamp, selected_row_ids, elements):
+    """ Update the elements of the cytoscape graph.
+
+    This function is called on startup, when the refresh button is clicked, and when a row is selected in the User Journey Datatable.
+    We need this callback to handle these (generally unrelated) situations because Dash only supports assigning
+    a single callback to a given output element.
+
+    Returns:
+        A dictionary of cytoscape elements describing the nodes and edges of the graph.
+    """
+
+    # regenerate elements on initial call and callbacks invoked by refresh button
+    regenerate_elements = True
+
+    ctx = dash.callback_context
+
+    if ctx.triggered:
+        # ctx.triggered[0] is either "cytoscape-graph.tapNode" or "client-dropdown.value"
+        triggered_id, triggered_prop = ctx.triggered[0]["prop_id"].split(".")
+        if triggered_id == "refresh-button":
+            state.clear_cache()
+        else:
+            # don't regenerate elements when triggered by a client datatable row selection
+            regenerate_elements = False
+
+    if regenerate_elements:
+        node_name_message_map, client_name_message_map = state.get_message_maps()
+
+        cytoscape_graph_elements = converters.cytoscape_elements_from_maps(
+            node_name_message_map,
+            client_name_message_map,
+        )
+    else:
+        if selected_row_ids != [None]:
+            user_journey_name = selected_row_ids[0][0]
+        else:
+            user_journey_name = None
+
+        cytoscape_graph_elements = transformers.apply_highlighted_edge_class_to_elements(
```

Comment:
These callbacks were previously in app.py, now refactored them to their own module.

Here's where the callback calls the helper function to apply the highlighting to relevant edges.